src/main/scala/com/cognite/sdk/scala/common/OAuth2.scala

```diff
@@ -0,0 +1,85 @@
+package com.cognite.sdk.scala.common
+
+import java.util.concurrent.TimeUnit
+
+import cats.Monad
+import cats.syntax.all._
+import cats.effect.{Clock, ConcurrentEffect}
+import com.cognite.sdk.scala.common.internal.{CachedResource, ConcurrentCachedObject}
+import com.softwaremill.sttp._
+import com.softwaremill.sttp.circe.asJson
+import io.circe.Decoder
+import io.circe.derivation.deriveDecoder
+
+object OAuth2 {
+  final case class ClientCredentials(
+    tokenUri: Uri,
+    clientId: String,
+    clientSecret: String,
+    scopes: List[String]
+  )
+
+  class ClientCredentialsProvider[F[_]] private (
+    cache: CachedResource[F, TokenState]
+  )(
+    implicit F: Monad[F],
+    clock: Clock[F],
+    sttpBackend: SttpBackend[F, Nothing]
+  ) extends AuthProvider[F] with Serializable {
+    def getAuth: F[Auth] = {
+      for {
+        now <- clock.monotonic(TimeUnit.SECONDS)
+        _ <- cache.invalidateIfNeeded(state => state.expiresAt <= now)
```

Comment:
The `=> _ <=` had me thinking for a while, hehe.

Reply:
Yeah, that's a bit puzzling. We could use an anonymous parameter (`_.expiresAt <= now`) or flip the comparison (`state => now > state.expiresAt`)

Reply:
It's funny that it can happen though, I don't mind keeping it either.